src/main/java/uk/nhs/digital/nhsconnect/lab/results/inbound/NhsAckProducerService.java

```diff
@@ -0,0 +1,159 @@
+package uk.nhs.digital.nhsconnect.lab.results.inbound;
+
+import com.github.mustachejava.Mustache;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import uk.nhs.digital.nhsconnect.lab.results.mesh.message.WorkflowId;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.Interchange;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.NhsAckContent;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.NhsAckStatus;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.NhsAckMessageContent;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.InterchangeHeader;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.MessageHeader;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.message.InterchangeParsingException;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.message.MessageParsingException;
+import uk.nhs.digital.nhsconnect.lab.results.sequence.SequenceService;
+import uk.nhs.digital.nhsconnect.lab.results.utils.TimestampService;
+import uk.nhs.digital.nhsconnect.lab.results.inbound.MessageProcessingResult.Success;
+import uk.nhs.digital.nhsconnect.lab.results.inbound.MessageProcessingResult.Error;
+
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.List;
+
+import static uk.nhs.digital.nhsconnect.lab.results.utils.TemplateUtils.loadTemplate;
+import static uk.nhs.digital.nhsconnect.lab.results.utils.TemplateUtils.fillTemplate;
+
+@Component
+@RequiredArgsConstructor(onConstructor_ = @Autowired)
+public class NhsAckProducerService {
+
+    @Autowired
+    private final TimestampService timestampService;
+
+    @Autowired
+    private final SequenceService sequenceService;
+
+    private static final int COMMON_SEGMENT_COUNT = 5;
```

Comment:
What does `COMMON_SEGMENT_COUNT` refer to & why is it hardwired to be 5?

Reply:
Depending on which status code we are returning with the NHSACK, there will be a different number of segments in the message. However, 5 of those segments appear in every type of NHSACK. Extra segments are added onto this in the relevant methods to give the total segment count for the message, so that it can be added into the trailer.

I've renamed this field to MESSAGE_COMMON_SEGMENT_COUNT in my changes, to make it clear that it's only the segments in the message that count, not the whole interchange.

If you can think of a better name/better way of doing this though, I'm happy to change it.